truffle/src/main/java/org/jruby/truffle/runtime/rope/ConcatRope.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2015, 2016 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle.runtime.rope;
+
+import org.jcodings.Encoding;
+import org.jruby.truffle.runtime.core.StringOperations;
+
+public class ConcatRope extends Rope {
+
+    private final Rope left;
+    private final Rope right;
+
+    private byte[] bytes;
+
+    public ConcatRope(Rope left, Rope right, Encoding encoding) {
+        super(encoding,
+                StringOperations.commonCodeRange(left.getCodeRange(), right.getCodeRange()),
+                left.isSingleByteOptimizable() && right.isSingleByteOptimizable(),
+                left.byteLength() + right.byteLength(),
+                left.characterLength() + right.characterLength(),
+                Math.max(left.depth(), right.depth()) + 1);
+
+        this.left = left;
+        this.right = right;
+    }
+
+    @Override
+    public byte[] getBytes() {
+        if (bytes == null) {
+            final byte[] leftBytes = left.getBytes();
+            final byte[] rightBytes = right.getBytes();
+
+            bytes = new byte[leftBytes.length + rightBytes.length];
+            System.arraycopy(leftBytes, 0, bytes, 0, leftBytes.length);
+            System.arraycopy(rightBytes, 0, bytes, leftBytes.length, rightBytes.length);
```

Comment:
For deep trees this is going to create a lot of intermediate arrays. Can't we allocate an array of the right size for the final string, and then pass it down into child nodes and get them to fill it in, rather than creating a new array each time?